src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -232,42 +262,182 @@ private void OnEntryEvicted(object key, object value, EvictionReason reason, obj
             }
         }
 
-        private async Task DisposeCircuitHost(CircuitHost circuitHost)
+        private async Task DisposeCircuitEntry(DisconnectedCircuitEntry entry)
         {
+            DisposeTokenSource(entry);
+
             try
             {
-                await circuitHost.DisposeAsync();
+                await entry.CircuitHost.DisposeAsync();
             }
             catch (Exception ex)
             {
                 Log.UnhandledExceptionDisposingCircuitHost(_logger, ex);
             }
         }
 
+        private void DisposeTokenSource(DisconnectedCircuitEntry entry)
+        {
+            try
+            {
+                entry.TokenSource.Dispose();
+            }
+            catch (Exception ex)
+            {
+                Log.ExceptionDisposingTokenSource(_logger, ex);
+            }
```

Comment:
This is interesting. For what reasons do you expect it's possible that `entry.TokenSource.Dispose()` would throw? Does this trigger a synchronous callback in our code somewhere?